akka-tools-persistence/src/main/scala/no/nextgentel/oss/akkatools/aggregate/GeneralAggregateBase.scala

```diff
@@ -48,6 +49,32 @@ abstract class GeneralAggregateBase[E:ClassTag, S <: AggregateStateBase[E, S]:Cl
   private val defaultErrorHandler = (errorMsg:String) => log.debug("No cmdFailed-handler executed")
 
 
+  protected override def onSnapshotOffer(offer : SnapshotOffer) : Unit = {
+    throw new Exception(s"Handling of snapshot recovery not defined, received ${offer.metadata}")
+  }
+
+  protected def acceptSnapshotRequest(request : SaveSnapshotOfCurrentState) : Boolean = {
+    log.info(s"Snapshot handling is not defined, rejecting $request")
+    false
+  }
+
+  protected def onSnapshotSuccess(success : SaveSnapshotSuccess) : Unit = {
+    log.info(s"Saved snapshot $success")

```

Comment:
Idea:

If SaveSnapshotOfCurrentState has property deleteMsgBeforeSnapshot, then you can delete it here.

and if you add to props to InternalState:
deleteUpToSeqNo (set it to current-seqNo in offerSnapshot if deleteMsgBeforeSnapshot)
hasDeleteUpToSeqNo.

The you have a method checkForMsgsToDelete() which you call in onSnapshotSuccess and onRecoverComplete().

It checks if hasDeleteUpToSeqNo < deleteUpToSeqNo. If it is, it send new internal cmd to "this".. which when received, will delete msgs and when done, add event which will update hasDeleteUpToSeqNo.

I think this will be a robust solution. It will in 99.9% cases do it right away, but if it does not work/crash, it will do it on next reload.

Just some thoughts.